### PR TITLE
feat(general): add Terraform JSON support

### LINKED ISSUE #4085: CDKTF Support
**Describe the issue**
Checkov currently supports AWS CDK, please can support be extended to CDKTF.

**Examples**
Extend checks that analyse AWS CDK for security misconfiguration, to cover CDKTF too.

**Version (please complete the following information):**
N/A

**Additional context**
N/A

### LINKED ISSUE #4283: Terraform(CDK) json
**CDK(Terraform)**
As I can see  `checkov` can't detect terraform json files that were synthesized by CDK. I use a workaround where I convert all json files to hcl, but unfortunately there is no official tool and  community tools aren't in active development phase and should be used only on your own risk .
https://github.com/kvz/json2hcl
https://github.com/relusc/hclconv

### PR DESCRIPTION
**Pull request by gruebel**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

- adds Terraform JSON syntax support
- it leverages most of the Terraform runner code and prepares the data to be processable by it
- inline skips for Python checks will follow in an other PR

Fixes #4085 
Fixes #4283

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform_json/parser.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        if err.problem in ("found character '\\t' that cannot start any token", "found unknown escape character"):
+            try:
+                result = json_parse(file_path, allow_nulls=False)
+                if result:
+                    template, template_lines = result  # type:ignore[assignment]  # this is handled by the next line
+                    if isinstance(template, list):
+                        # should not happen and is more relevant for type safety
+                        template = template[0]
+            except Exception:
+                logger.error(f"Template {file_path} is malformed: {err.problem}")
```
Comment (bo156) [inline checkov/terraform_json/parser.py:50]: do you want to also log the second error? since here you basically only log the original one

File: checkov/terraform_json/parser.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    except YAMLError:
+        pass
+
+    if template is None or template_lines is None:
+        return None, None
+
+    return template, template_lines
+
+
+def load(file_path: Path) -> tuple[dict[str, Any], list[tuple[int, str]]]:
```
Comment (bo156) [inline checkov/terraform_json/parser.py:61]: 2 questions here:
1. why is this needed instead of just using `json.load`?
2. It's name is confusing as I expected it to be `json.load`, so if it shouldn't be like that I would expect the name to reflect what is special about this load function. Also it is a saved word in `ipython` which added to my confusion as I initially thought this is a general function 😅

File: checkov/terraform_json/parser.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    parent_key: str | None = None,
+) -> dict[str, list[Any]]:
+    ret_dict = {}
+
+    if not isinstance(obj, dict):
+        raise Exception("this method receives only dicts")
+
+    for key, value in obj.items():
+        if key == COMMENT_FIELD_NAME:
+            # don't hclify the comment block
```
Comment (bo156) [inline checkov/terraform_json/parser.py:127]: did you just coin a new word? 🤣

File: checkov/terraform_json/runner.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            report.add_parsing_errors(parsing_errors)
+
+            if external_checks_dir:
+                for directory in external_checks_dir:
+                    resource_registry.load_external_checks(directory)
+
+                    if CHECKOV_CREATE_GRAPH:
+                        self.graph_registry.load_external_checks(directory)
+
+            # TODO: create function 'build_definitions_context()'
```
Comment (bo156) [inline checkov/terraform_json/runner.py:91]: should this comments stay here? If yes, is there a linked issue for it (or do we want to create one)?

File: checkov/terraform_json/runner.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+if TYPE_CHECKING:
+    from checkov.common.graph.checks_infra.registry import BaseRegistry
+    from checkov.terraform.graph_manager import TerraformGraphManager
+    from checkov.common.typing import LibraryGraphConnector, _CheckResult
+
+logger = logging.getLogger(__name__)
+
+
+class Runner(TerraformRunner):
```
Comment (bo156) [inline checkov/terraform_json/runner.py:31]: Should probably be called `TerraformJsonRunner` to avoid multiple instances of the name `Runner` in the code

**Review by bo156 — CHANGES_REQUESTED** [review]

File: checkov/terraform_json/parser.py (hunk 2 above)
Comment (gruebel) [reply checkov/terraform_json/parser.py:61]: we need line numbers in the returned `dict`, which normal `json.load(s)` can't offer. Sure, I will rename it, it is somehow named that way in other parsers, so I didn't think much about it 😄

File: checkov/terraform_json/parser.py (hunk 1 above)
Comment (gruebel) [reply checkov/terraform_json/parser.py:50]: the line below logs the whole stack trace, which is more than enough for me 🙂

File: checkov/terraform_json/parser.py (hunk 3 above)
Comment (gruebel) [reply checkov/terraform_json/parser.py:127]: sounds like 😄 but no, we use the name in the Terraform plan parser

File: checkov/terraform_json/runner.py (hunk 4 above)
Comment (gruebel) [reply checkov/terraform_json/runner.py:91]: #4635 4635

**Comment by mpicard** [comment]
Is this documented? I am trying to use checkov with cdktf and I am not sure it works or if i did not configure checkov correctly

**Comment by gruebel** [comment]
hey @mpicard yeah I didn't add proper docs yet. Wanted to do it, when I implement the inline suppression feature.
You need to synthesize your stack and then you can run `checkov` against it.
ex.
```shell
cdktf synth
checkov -d cdktf.out
```

File: checkov/terraform_json/parser.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+def loads(file_path: Path) -> tuple[dict[str, Any], list[tuple[int, str]]]:
+    """Loads the given JSON file with line numbers"""
+
+    try:
+        content = file_path.read_text()
+    except UnicodeDecodeError:
+        logging.debug(f"Encoding for file {file_path} is not UTF-8, trying to detect it")
+        content = str(from_path(file_path).best())
+
+    if not all(key in content for key in ("resource", "provider")):
```
Comment (lirshindalman) [inline checkov/terraform_json/parser.py:70]: hey, why did you add this condition? 
In Terraform we don't have this limitation in tf file and it works ok without this, like tf file without them

Comment (gruebel) [reply checkov/terraform_json/parser.py:70]: If I remember correctly, then I added this condition to be able to differentiate between a random JSON file and a Terraform generated JSON file, otherwise you can get strange errors for trying to process a random JSON file.
